### ISSUE TITLE
U-Net architecture for score-based generative models

### ISSUE THREAD

**Issue by anh-tong** [patrick-kidger/equinox#211]
Hi Patrick,

The current backbone to approximate score functions in [`examples/score_based_diffusion.ipynb`](https://github.com/patrick-kidger/equinox/blob/main/examples/score_based_diffusion.ipynb) uses Mixer (I guess, due to the simplicity of this model). I observe that this takes quite long time to run (using 10^6 iterations during optimization). As noted in [Yang Song's blog](https://yang-song.net/blog/2021/score/), U-Net is the key architecture to have a good quality of image generations. 

I have tried to implement U-Net using Equinox (quite easy to translate from PyTorch to Equinox) based on [huggingface/diffusers](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d.py) in this file
https://github.com/anh-tong/equinox/blob/unet/examples/unet.py
An output can be found here
https://github.com/anh-tong/equinox/blob/unet/examples/score_based_diffusion.nbconvert.ipynb
It seems this implementation converges faster than that of Mixer (at step 2*10^3, the loss is already roughly similiar to Mixer at 10^4).
Overall training time is around 6-7 hours.

I wonder if you want to add U-Net as an alternative architecture here. However, the implementation is not as clean and succtint as Mixer.

**Comment by patrick-kidger** [comment]
Right! So I chose Mixer here just because it was simple, and didn't take up too much space on the page. In contrast U-net implementations are always ginormous.

Cc @paganpasta : is something like this something that would make sense to add to eqxvision? I've got my own (so far private) UNet implementation around that I could probably contribute as well.

**Comment by paganpasta** [comment]
@patrick-kidger : Yeah, I think it makes sense to add U-net to eqxvision. I have also seen variations of U-nets been applied to Image Segmentation as well. 

I'd be happy to accept a PR on this.
